src/schema/date_time_options.rs

```diff
@@ -0,0 +1,606 @@
+    fn from(_: ()) -> DateTimeOptions {
+        DateTimeOptions::default()
+    }
+}
+
+impl From<FastFlag> for DateTimeOptions {
+    fn from(_: FastFlag) -> Self {
+        DateTimeOptions {
+            indexed: false,
+            fieldnorms: false,
+            stored: false,
+            fast: Some(Cardinality::SingleValue),
+            ..Default::default()
+        }
+    }
+}
+
+impl From<StoredFlag> for DateTimeOptions {
+    fn from(_: StoredFlag) -> Self {
+        DateTimeOptions {
+            indexed: false,
+            fieldnorms: false,
+            stored: true,
+            fast: None,
+            ..Default::default()
+        }
+    }
+}
+
+impl From<IndexedFlag> for DateTimeOptions {
+    fn from(_: IndexedFlag) -> Self {
+        DateTimeOptions {
+            indexed: true,
+            fieldnorms: true,
+            stored: false,
+            fast: None,
+            ..Default::default()
+        }
+    }
+}
+
+impl<T: Into<DateTimeOptions>> BitOr<T> for DateTimeOptions {
+    type Output = DateTimeOptions;
+
+    fn bitor(self, other: T) -> DateTimeOptions {
+        let other = other.into();
+        DateTimeOptions {
+            indexed: self.indexed | other.indexed,
+            fieldnorms: self.fieldnorms | other.fieldnorms,
+            stored: self.stored | other.stored,
+            fast: self.fast.or(other.fast),
+            // TODO: Is the default ok for the other field?
+            ..Default::default()
+        }
+    }
+}
+
+impl<Head, Tail> From<SchemaFlagList<Head, Tail>> for DateTimeOptions
+where
+    Head: Clone,
+    Tail: Clone,
+    Self: BitOr<Output = Self> + From<Head> + From<Tail>,
+{
+    fn from(head_tail: SchemaFlagList<Head, Tail>) -> Self {
+        Self::from(head_tail.head) | Self::from(head_tail.tail)
+    }
+}
+
+// Parser function types
+type StringDateTimeParser = Arc<dyn Fn(&str) -> Result<OffsetDateTime, String> + Sync + Send>;
+type NumberDateTimeParser = Arc<dyn Fn(i64) -> Result<OffsetDateTime, String> + Sync + Send>;
+
+/// A struct for holding datetime parsing functions.
+#[derive(Clone)]
+pub struct DateTimeParsersHolder {
+    /// Functions for parsing string input as datetime.
+    string_parsers: VecDeque<StringDateTimeParser>,
+    /// Function for parsing number input as timestamp.
+    number_parser: NumberDateTimeParser,
+}
+
+impl Default for DateTimeParsersHolder {
+    fn default() -> Self {
+        Self {
+            string_parsers: VecDeque::new(),
+            number_parser: make_timestamp_parser(DateTimePrecision::Milliseconds),
+        }
+    }
+}
+
+impl DateTimeParsersHolder {
+    /// Parses string input.
+    pub fn parse_string(&mut self, value: String) -> Result<OffsetDateTime, String> {
+        for (index, parser) in self.string_parsers.iter().enumerate() {
+            if let Ok(date_time) = parser(&value) {
+                // Move successful parser in front of queue.
+                self.string_parsers.swap(0, index);
```

Comment:
Ah I see... 
I think Vec has swap too.

I am not sure this is a good idea.

So with the HashSet defintion, the output of quickwit right now is not really deterministic at compile time.
Date format can be ambiguous ( think about MM-DD-YYY vs DD-MM-YYYY  and 01/03/2021).
Which means that our output depends on the order of our date formats.

By using a HashSet, you make the output of quickwit dependent on the implementation of hashset.
(A change in the rust standard library fro instance, would break unit tests.) 
We need to stay away from HashSet.

I would also stay away from the reordering logic, but I wonder...
So it has two functions I believe:
- a) performance.
- b) best effort to deal with the ambiguity described above.

a) can be partially addressed by extracting ez to check predicate for each format.
For instance on the date string length.
- If the date is not exactly 10chrs. skip format XXXXX.

b) is tricky. With your solution we can "eventually" find the right format when we reach a date that is unambiguous (e.g. 24/02/2021). On the other hand, I think we should give all the tools to the user to tell use the right thing. I'd rather not do the swapping logic and just rely on the priority logic.